Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-239.ca-central-1.compute.internal
Sha: fdd18b279e807866678bddbebf8e1bdbef248c12
Number of Error(s): 19
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
CMake Error at /home/centos/smoketest/PR-14840/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 502 Bad Gateway
make[2]: *** [CMakeFiles/generate-boost-regex.dir/build.make:90: generate-boost-regex-prefix/src/generate-boost-regex-stamp/generate-boost-regex-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:937: CMakeFiles/generate-boost-regex.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
CMake Error at /home/centos/smoketest/PR-14840/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
  Each download failed!
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
         status_code: 22
         status_string: HTTP response code said error
         log:
         --- LOG BEGIN ---
           Trying 34.215.63.224...
  TCP_NODELAY set
  Connected to dl.bintray.com (34.215.63.224) port 443 (#0)
  ALPN, offering http/1.1
  successfully set certificate verify locations:
    CAfile: /etc/pki/tls/certs/ca-bundle.crt
    CApath: none
  [5 bytes data]
  TLSv1.3 (OUT), TLS handshake, Client hello (1):
  [512 bytes data]
  [5 bytes data]
  TLSv1.3 (IN), TLS handshake, Server hello (2):
  [102 bytes data]
  NPN, negotiated HTTP1.1
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Certificate (11):
  [2765 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server key exchange (12):
  [333 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server finished (14):
  [4 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
  [70 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
  [1 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Next protocol (67):
  [36 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Finished (20):
  [16 bytes data]
  [5 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Finished (20):
  [16 bytes data]
  SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
  ALPN, server did not agree to a protocol
  Server certificate:
   subject: CN=*.bintray.com
   start date: Sep 26 00:00:00 2019 GMT
   expire date: Nov  9 12:00:00 2021 GMT
   subjectAltName: host dl.bintray.com matched cert\s *.bintray.com
   issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 201
 ... ( comment is too long, 1188 lines, 100636 bytes, truncated) ... 
[Build error:](https://gist.github.com/e9db016b3451a27ab63244dd0090de2d)
Build log [Build log](https://gist.github.com/HPCCSmoketest/e9db016b3451a27ab63244dd0090de2d#file-result-21-04-12-13-42-20-log)

```

--- a/dali/dfu/dfurun.cpp
+++ b/dali/dfu/dfurun.cpp
@@ -1314,6 +1314,16 @@ public:
                                 };
                             }
                         }
+#ifdef _CONTAINERIZED
+                        StringBuffer clusterName;
+                        destination->getGroupName(0, clusterName);
+                        Owned<IPropertyTree> plane = getDropZonePlane(clusterName);
+                        if (plane)
+                        {
+                            if (plane->hasProp("@defaultSprayParts"))
+                                destination->setNumPartsOverride(plane->getPropInt("@defaultSprayParts"));
+                        }
+#endif
                         if (destination->getWrap())
                         {
                             Owned<IFileDescriptor> fdesc = source?source->getFileDescriptor():NULL;

--- a/esp/services/ws_fs/ws_fsService.cpp
+++ b/esp/services/ws_fs/ws_fsService.cpp
@@ -1870,6 +1870,7 @@ bool CFileSprayEx::onSprayFixed(IEspContext &context, IEspSprayFixed &req, IEspS
 
         wu->setJobName(destTitle.str());
         const char * dfuQueue = req.getDFUServerQueue();
+#ifndef _CONTAINERIZED
         Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(true);
         Owned<IConstEnvironment> constEnv = envFactory->openEnvironment();
         if (!isEmptyString(dfuQueue))
@@ -1877,6 +1878,24 @@ bool CFileSprayEx::onSprayFixed(IEspContext &context, IEspSprayFixed &req, IEspS
             if (!constEnv->isValidDfuQueueName(dfuQueue))
                 throw MakeStringException(ECLWATCH_INVALID_INPUT, "invalid DFU server queue name:'%s'", dfuQueue);
         }
+#else
+        bool isValidDfuQueueName = false;
+        Owned<IPropertyTreeIterator> dfuServers = queryComponentConfig().getElements("dfuQueues");
+        ForEach(*dfuServers)
+        {
+            IPropertyTree & dfuServer = dfuServers->query();
+            const char * dfuServerName = dfuServer.queryProp("@name");
+            StringBuffer knownDfuQueueName;
+            getDfuQueueName(knownDfuQueueName, dfuServerName);
+            if (streq(dfuQueue, knownDfuQueueName))
+            {
+                isValidDfuQueueName = true;
+                break;
+            }
+        }
+        if (!isValidDfuQueueName)
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid DFU server queue name: '%s'", dfuQueue);
+#endif
         setDFUServerQueueReq(dfuQueue, wu);
         setUserAuth(context, wu);
         wu->setCommand(DFUcmd_import);
@@ -3434,7 +3453,17 @@ bool CFileSprayEx::onGetSprayTargets(IEspContext &context, IEspGetSprayTargetsRe
     try
     {
         context.ensureFeatureAccess(FILE_SPRAY_URL, SecAccess_Read, ECLWATCH_FILE_SPRAY_ACCESS_DENIED, "Permission denied.");
-
+#ifdef _CONTAINERIZED
+        IArrayOf<IEspGroupNode> sprayTargets;
+        Owned<IPropertyTreeIterator> dataPlanes = queryGlobalConfig().getElements("storage/planes[labels='data']");
+        ForEach(*dataPlanes)
+        {
+            IPropertyTree & plane = dataPlanes->query();
+            const char * name = plane.queryProp("@name");
+            appendGroupNode(sprayTargets, name, "storage plane", false /*replicate outputs*/);
+        }
+        resp.setGroupNodes(sprayTargets);
+#else
         Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
         Owned<IConstEnvironment> environment = factory->openEnvironment();
         Owned<IPropertyTree> root = &environment->getPTree();
@@ -3488,6 +3517,7 @@ bool CFileSprayEx::onGetSprayTargets(IEspContext &context, IEspGetSprayTargetsRe
         }
 
         resp.setGroupNodes(sprayTargets);
+#endif
     }
     catch(IException* e)
     {

--- a/esp/services/ws_topology/ws_topologyService.cpp
+++ b/esp/services/ws_topology/ws_topologyService.cpp
@@ -1450,28 +1450,29 @@ bool CWsTopologyEx::onTpServiceQuery(IEspContext &context, IEspTpServiceQueryReq
         if (!type || !*type || (strcmp(eqAllServices,type) == 0))
         {
             IEspTpServices& ServiceList = resp.updateServiceList();
-
+            m_TpWrapper.getTpDfuServers( ServiceList.getTpDfuServers() );
+            m_TpWrapper.getTpDropZones(version, nullptr, true, ServiceList.getTpDropZones() );
+#ifndef _CONTAINERIZED
             m_TpWrapper.getTpDaliServers( version, ServiceList.getTpDalis() );
             m_TpWrapper.getTpEclServers( ServiceList.getTpEclServers() );
             m_TpWrapper.getTpEclCCServers( ServiceList.getTpEclCCServers() );
             m_TpWrapper.getTpEclAgents( ServiceList.getTpEclAgents() );
             m_TpWrapper.getTpEspServers( ServiceList.getTpEspServers() );
-            m_TpWrapper.getTpDfuServers( ServiceList.getTpDfuServers() );   
-            m_TpWrapper.getTpSashaServers( ServiceList.getTpSashaServers() );   
+            m_TpWrapper.getTpSashaServers( ServiceList.getTpSashaServers() );
             m_TpWrapper.getTpGenesisServers( ServiceList.getTpGenesisServers() );
             m_TpWrapper.getTpLdapServers( ServiceList.getTpLdapServers() );
-            m_TpWrapper.getTpDropZones(version, nullptr, true, ServiceList.getTpDropZones() );
             m_TpWrapper.getTpFTSlaves( ServiceList.getTpFTSlaves() );
             m_TpWrapper.getTpDkcSlaves( ServiceList.getTpDkcSlaves() );
 
             if (version > 1.15)
-            {       
+            {
                 m_TpWrapper.getTpEclSchedulers( ServiceList.getTpEclSchedulers() );
             }
             if (version >= 1.28)
-            {       
+            {
                 m_TpWrapper.getTpSparkThors(version, nullptr, ServiceList.getTpSparkThors() );
             }
+#endif
         }
 
         resp.setMemThreshold( m_memThreshold );
@@ -1481,15 +1482,15 @@ bool CWsTopologyEx::onTpServiceQuery(IEspContext &context, IEspTpServiceQueryReq
         resp.setDiskThresholdType( m_bDiskThresholdIsPercentage ? "0" : "1");
 
         if (version > 1.06 && m_bEncapsulatedSystem)
-        {       
+        {
             resp.setEncapsulatedSystem( m_bEncapsulatedSystem );
         }
         if (version > 1.07)
-        {       
+        {
             resp.setEnableSNMP(m_enableSNMP);
         }
         if ((version > 1.12) && (m_preflightProcessFilter.length() > 0))
-        {       
+        {
             resp.setPreflightProcessFilter(m_preflightProcessFilter);
         }
         if (version >= 1.20)

--- a/esp/smc/SMCLib/TpWrapper.cpp
+++ b/esp/smc/SMCLib/TpWrapper.cpp
@@ -157,7 +157,7 @@ void CTpWrapper::fetchInstances(const char* ServiceType, IPropertyTree& service,
 
             IEspTpMachine* machine = createTpMachine("", "");
             getMachineInfo(*machine, instanceNode, "/Environment/Software", ServiceType, "@computer");
-         machine->setPort( instanceNode.getPropInt("@port") );
+            machine->setPort( instanceNode.getPropInt("@port") );
             const char* directory = instanceNode.queryProp("@directory");
             if (directory && *directory)
                 machine->setDirectory( directory );
@@ -464,8 +464,48 @@ void CTpWrapper::getTpEspServers(IArrayOf<IConstTpEspServer>& list)
     }
 }
 
+static IEspTpMachine * createLocalHostTpMachine(const char *path)
+{
+    Owned<IEspTpMachine> machine = createTpMachine();
+    IpAddress ipAddr;
+    ipAddr.ipset("localhost");
+    StringBuffer localHost;
+    ipAddr.getIpText(localHost);
+    machine->setName(localHost.str());
+    machine->setNetaddress(localHost.str());
+    machine->setConfigNetaddress("localhost");
+    machine->setDirectory(path);
+    machine->setOS(getPathSepChar(path) == '/' ? MachineOsLinux : MachineOsW2K);
+    return machine.getClear();
+}
+
 void CTpWrapper::getTpDfuServers(IArrayOf<IConstTpDfuServer>& list)
 {
+#ifdef _CONTAINERIZED
+    Owned<IPropertyTreeIterator> dfuQueues = queryComponentConfig().getElements("dfuQueues");
+    ForEach(*dfuQueues)
+    {
+        IPropertyTree & dfuQueue = dfuQueues->query();
+        const char * dfuName = dfuQueue.queryProp("@name");
+        StringBuffer queue;
+        getDfuQueueName(queue, dfuName);
+        Owned<IEspTpDfuServer> pService = createTpDfuServer("","");
+        pService->setName(dfuName);
+        pService->setDescription(dfuName);
+        pService->setBuild("");
+        pService->setQueue(queue);
+        pService->setType(eqDfu);
+        IArrayOf<IEspTpMachine> tpMachines;
+        Owned<IPropertyTreeIterator> planes = getDropZonePlanesIterator();
+        ForEach(*planes)
+        {
+            IPropertyTree & plane = planes->query();
+            tpMachines.append(*createLocalHostTpMachine(plane.queryProp("@prefix")));
+        }
+        pService->setTpMachines(tpMachines);
+        list.append(*pService.getClear());
+    }
+#else
     Owned<IPropertyTree> root = getEnvironment("Software");
     if (!root)
         throw MakeStringExceptionDirect(ECLWATCH_CANNOT_GET_ENV_INFO, MSG_FAILED_GET_ENVIRONMENT_INFO);
@@ -497,8 +537,9 @@ void CTpWrapper::getTpDfuServers(IArrayOf<IConstTpDfuServer>& list)
         fetchInstances(eqDfu, serviceTree, tpMachines);
         pService->setTpMachines(tpMachines);
 
-        list.append(*pService.getLink());
+        list.append(*pService.getClear());
     }
+#endif
 }
 
 
@@ -1681,27 +1722,17 @@ void CTpWrapper::getTpDropZones(double clientVersion, const char* name, bool ECL
     {
         IPropertyTree & plane = planes->query();
         const char * dropzonename = plane.queryProp("@name");
-        const char * prefix = plane.queryProp("@prefix");
-        StringBuffer path(prefix);
+        const char * path = plane.queryProp("@prefix");
         Owned<IEspTpDropZone> dropZone = createTpDropZone();
         dropZone->setName(dropzonename);
-        dropZone->setDescription(""); //TODO
+        dropZone->setDescription("");
         dropZone->setPath(path);
         dropZone->setBuild("");
-        dropZone->setECLWatchVisible(plane.getPropBool("@eclWatchVisible", true));
-        StringBuffer localHost;
+        dropZone->setECLWatchVisible(true);
         IArrayOf<IEspTpMachine> tpMachines;
-        Owned<IEspTpMachine> machine = createTpMachine();
-        IpAddress ipAddr;
-        ipAddr.ipset("localhost");
-        ipAddr.getIpText(localHost);
-        machine->setNetaddress(localHost.str());
-        machine->setConfigNetaddress("localhost");
-        machine->setDirectory(path);
-        machine->setOS(getPathSepChar(path) == '/' ? MachineOsLinux : MachineOsW2K);
-        tpMachines.append(*machine.getLink());
+        tpMachines.append(*createLocalHostTpMachine(path));
         dropZone->setTpMachines(tpMachines);
-        list.append(*dropZone.getLink());
+        list.append(*dropZone.getClear());
     }
 #else
     Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(true);

--- a/helm/hpcc/values.schema.json
+++ b/helm/hpcc/values.schema.json
@@ -412,7 +412,11 @@
         "umask" : {
           "description": "file creation mask (used by despray)",
           "type" : "string"
-        }
+        },
+        "defaultSprayParts" : {
+            "description": "Number of parts sprayed by default",
+            "type" : "integer"
+          }
       },
       "required": [ "name", "prefix" ],
       "additionalProperties": false

--- a/helm/hpcc/values.yaml
+++ b/helm/hpcc/values.yaml
@@ -95,6 +95,7 @@ storage:
   #   includeDeviceInPath: false            # Is the device number appended to the mount for the physical path for a file?  (Only required in unusual situations)
   #   hosts: <name>                         # Name of the host group for bare metal - must match the name of the storage plane..
   #   secret: <secret-id>                   # what secret is required to access the files.  This could optionally become a list if required (or add secrets:).
+  #   defaultSprayParts: 4                  # The number of partitions created when spraying (default: 1)
   #   options:                              # not sure if it is needed
 
   dllStorage:

